# Galois Field Demonstrations

This notebook demonstrates arithmetic in finite fields $GF(p^m)$, following Chapter 5 of the course notes.

We will:
- Construct $GF(p^m)$ using an irreducible (primitive) polynomial.
- Perform addition and multiplication in the field.
- List all elements of the field in both integer and polynomial (list) representation.
- Compute powers of a primitive element.
- Verify examples from the notes.

## Setup and Imports

We import the `GF` class, which implements arithmetic in $GF(p^m)$, and set up the Python environment.

In [1]:
import numpy as np
import sys
import os
# Ensure src is in the Python path for module imports
SRC_PATH = os.path.abspath(os.path.join(os.getcwd(), '../src'))
if SRC_PATH not in sys.path:
    sys.path.insert(0, SRC_PATH)
from codes.gf import GF

## Constructing $GF(8)$

We construct the field $GF(8)$ using the primitive polynomial $x^3 + x + 1$. In binary, this polynomial is represented as `[1, 1, 0, 1]` (corresponding to $1 + x + x^3$).

The field has $2^3 = 8$ elements.

In [2]:
# Example: GF(8) with primitive polynomial x^3 + x + 1
gf8 = GF(2, 3, [1, 1, 0, 1])  # binary polynomial: 1 + x + x^3

## Table of All Elements in $GF(8)$

Below, each element of $GF(8)$ is shown in three forms:
- **Binary representation** (as a 3-bit string)
- **Power of the primitive element $\alpha$** (with $\alpha = x$), or `0` for the zero element
- **Polynomial (list) representation**

In [3]:
gf8.print_elements_table()

  Binary | Alpha Power |      Polynomial
----------------------------------------
     001 |        α^0 |               1
     010 |        α^1 |               α
     100 |        α^2 |             α^2
     011 |        α^3 |           1 + α
     110 |        α^4 |         α + α^2
     111 |        α^5 |     1 + α + α^2
     101 |        α^6 |         1 + α^2


## Addition Example

We demonstrate addition in $GF(8)$ by adding $\alpha$ and $\alpha^2$. The result is $\alpha + \alpha^2$.

In [4]:
a = [0,1,0]
b = [0,0,1]

result = gf8.add(a, b)  # Should return [0, 1, 1] (binary addition)
print(f"Addition result: {result}")


Addition result: 6


Note that the result from the addition is the integer 6. Note in the table above that $\alpha$ is $010$ in binary representation and $\alpha^2$ is $100$ in binary reprentation. The sum of this in binary is $110$ which is the same as the integer 6. We can print this result in polynomial format using the `poly_str` mehtod.

In [5]:
print(gf8.poly_str(result))

α + α^2


## Multiplication Example

We demonstrate multiplication in $GF(8)$ by multiplying $x$ and $x^2$. The result is $x^3$, which is reduced modulo the primitive polynomial $x^3 + x + 1$.

In [6]:
prod_result = gf8.mul(a, b)
print("\u03B1 * \u03B1^2 =", gf8.poly_str(prod_result))

α * α^2 = 1 + α
